File: server/common/src/main/java/su/plo/voice/server/audio/source/VoiceServerEntitySource.java
package su.plo.voice.server.audio.source;

import lombok.Getter;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import su.plo.lib.api.server.entity.MinecraftServerEntity;
import su.plo.lib.api.server.world.ServerPos3d;
import su.plo.voice.api.addon.AddonContainer;
import su.plo.voice.api.server.PlasmoVoiceServer;
import su.plo.voice.api.server.audio.line.ServerSourceLine;
import su.plo.voice.api.server.audio.source.ServerEntitySource;
import su.plo.voice.proto.data.audio.source.EntitySourceInfo;

public final class VoiceServerEntitySource extends BaseServerSource<EntitySourceInfo> implements ServerEntitySource {

    @Getter
    private final MinecraftServerEntity entity;
    private final ServerPos3d playerPosition = new ServerPos3d();

    public VoiceServerEntitySource(@NotNull PlasmoVoiceServer voiceServer,
                                   @NotNull AddonContainer addon,
                                   @NotNull ServerSourceLine line,
                                   @Nullable String codec,
                                   boolean stereo,
                                   @NotNull MinecraftServerEntity entity) {
        super(voiceServer, addon, entity.getUUID(), line, codec, stereo);

        this.entity = entity;
    }

    @Override
    public @NotNull ServerPos3d getPosition() {
        return entity.getServerPosition(playerPosition);
    }

    @Override
    public @NotNull EntitySourceInfo getInfo() {
        return new EntitySourceInfo(
                addon.getId(),
                id,
                line.getId(),
                (byte) state.get(),
                codec,
                stereo,
                iconVisible,
                angle,
                entity.getId()
        );
    }
}

File: server/common/src/main/kotlin/su/plo/voice/server/audio/source/VoiceServerEntitySource.kt
package su.plo.voice.server.audio.source


import su.plo.slib.api.server.entity.McServerEntity
import su.plo.slib.api.server.position.ServerPos3d
import su.plo.voice.api.addon.AddonContainer
import su.plo.voice.api.server.PlasmoVoiceServer
import su.plo.voice.api.server.audio.line.ServerSourceLine
import su.plo.voice.api.server.audio.source.ServerEntitySource
import su.plo.voice.proto.data.audio.codec.CodecInfo
import su.plo.voice.proto.data.audio.source.EntitySourceInfo
import java.util.UUID
import java.util.concurrent.CompletableFuture

class VoiceServerEntitySource(
    private val voiceServer: PlasmoVoiceServer,
    addon: AddonContainer,
    line: ServerSourceLine,
    decoderInfo: CodecInfo?,
    stereo: Boolean,
    override val entity: McServerEntity
) : VoiceServerProximitySource<EntitySourceInfo>(voiceServer, addon, UUID.randomUUID(), line, decoderInfo, stereo),
    ServerEntitySource {

    override val position: ServerPos3d
        get() = entity.getServerPosition()

    override val sourceInfo: EntitySourceInfo
        get() = EntitySourceInfo(
            addon.id,
            id,
            line.id,
            name,
            state.get().toByte(),
            decoderInfo,
            stereo,
            iconVisible,
            angle,
            entity.id
        )

    override fun resolveSourceInfo(): CompletableFuture<EntitySourceInfo> =
        voiceServer.minecraftServer.scheduler.runTaskFor(entity) { sourceInfo }
}
